sklearn/linear_model/tests/test_ridge.py

```diff
@@ -938,3 +938,31 @@ def test_dtype_match_cholesky():
     assert ridge_32.predict(X_32).dtype == X_32.dtype
     assert ridge_64.predict(X_64).dtype == X_64.dtype
     assert_almost_equal(ridge_32.coef_, ridge_64.coef_, decimal=5)
+
+
+@pytest.mark.parametrize(
+    'solver', ['svd', 'cholesky', 'lsqr', 'sparse_cg', 'sag', 'saga'])
+@pytest.mark.parametrize('assert_tolerance', [1e-1, 1e-2, 1e-3, 1e-6])
+def test_ridge_regression_dtype_stability(solver, assert_tolerance):
+    n_samples, n_features = 6, 5
+    X = rng.randn(n_samples, n_features)
+    y = rng.randn(n_samples)
+    RANDOM_STATE = np.random.RandomState(0)
+    ALPHA = 1.0
+
+    results = { current_dtype: ridge_regression(X.astype(current_dtype),
+                                                y.astype(current_dtype),
+                                                alpha=ALPHA,
+                                                solver=solver,
+                                                random_state=RANDOM_STATE,
+                                                sample_weight=None,
+                                                max_iter=None,
+                                                tol=1e-3,
+                                                return_n_iter=False,
+                                                return_intercept=False)
+                for current_dtype in (np.float32, np.float64)}
+
+    assert results[np.float32].dtype == np.float32
+    assert results[np.float64].dtype == np.float64
+    assert_allclose(results[np.float32], results[np.float64],
+                    rtol=assert_tolerance)
```

Comment:
The types are maintained but the numerical results are not stable. See:

```sh
~/code/scikit-learn is/ridge_regression_sag*
(mne) ❯ pytest sklearn/linear_model/tests/test_ridge.py -k dtype_stability -v --tb=no      
Test session starts (platform: linux, Python 3.6.6, pytest 4.0.0, pytest-sugar 0.9.2)
cachedir: .pytest_cache
rootdir: /home/sik/code/scikit-learn, inifile: setup.cfg
plugins: sugar-0.9.2, pudb-0.7.0, mock-1.10.0, faulthandler-1.5.0, cov-2.6.0
collecting ... 
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.1-svd] ✓                       4% ▌         
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.1-cholesky] ✓                  8% ▉         
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.1-lsqr] ✓                     12% █▍        
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.1-sparse_cg] ✓                17% █▋        
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.1-sag] ✓                      21% ██▏       
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.1-saga] ✓                     25% ██▌       
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.01-svd] ✓                     29% ██▉       
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.01-cholesky] ✓                33% ███▍      
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.01-lsqr] ✓                    38% ███▊      
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.01-sparse_cg] ✓               42% ████▎     

 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.01-sag] ⨯                     46% ████▋     
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.01-saga] ✓                    50% █████     
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.001-svd] ✓                    54% █████▌    
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.001-cholesky] ✓               58% █████▉    
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.001-lsqr] ✓                   62% ██████▍   
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.001-sparse_cg] ✓              67% ██████▋   

 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.001-sag] ⨯                    71% ███████▏  

 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[0.001-saga] ⨯                   75% ███████▌  
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[1e-06-svd] ✓                    79% ███████▉  
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[1e-06-cholesky] ✓               83% ████████▍ 

 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[1e-06-lsqr] ⨯                   88% ████████▊ 
 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[1e-06-sparse_cg] ✓              92% █████████▎

 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[1e-06-sag] ⨯                    96% █████████▋

 sklearn/linear_model/tests/test_ridge.py::test_ridge_regression_dtype_stability[1e-06-saga] ⨯                  100% ██████████

Results (0.66s):
      18 passed
       6 failed
         - sklearn/linear_model/tests/test_ridge.py:943 test_ridge_regression_dtype_stability[0.01-sag]
         - sklearn/linear_model/tests/test_ridge.py:943 test_ridge_regression_dtype_stability[0.001-sag]
         - sklearn/linear_model/tests/test_ridge.py:943 test_ridge_regression_dtype_stability[0.001-saga]
         - sklearn/linear_model/tests/test_ridge.py:943 test_ridge_regression_dtype_stability[1e-06-lsqr]
         - sklearn/linear_model/tests/test_ridge.py:943 test_ridge_regression_dtype_stability[1e-06-sag]
         - sklearn/linear_model/tests/test_ridge.py:943 test_ridge_regression_dtype_stability[1e-06-saga]
      49 deselected
```

Similar issues can be found #13303 and #13309

Reply:
what is weird compared to #13309 is that `svd` is stable